Student/Student/Submissions/SubmissionButton/SubmissionButtonPresenter.swift

```diff
@@ -0,0 +1,271 @@
+    }
+
+    func cancel(_ controller: AudioRecorderViewController) {
+        controller.dismiss(animated: true, completion: nil)
+    }
+
+    func send(_ controller: AudioRecorderViewController, url: URL) {
+        controller.dismiss(animated: true) {
+            self.submitMediaType(.audio, url: url)
+        }
+    }
+
+    func imagePickerController(_ picker: UIImagePickerController, didFinishPickingMediaWithInfo info: [UIImagePickerController.InfoKey: Any]) {
+        picker.dismiss(animated: true) {
+            do {
+                if let videoURL = info[.mediaURL] as? URL {
+                    let destination = URL
+                        .temporaryDirectory
+                        .appendingPathComponent("videos", isDirectory: true)
+                        .appendingPathComponent(String(Clock.now.timeIntervalSince1970))
+                        .appendingPathExtension(videoURL.pathExtension)
+                    try videoURL.move(to: destination)
+                    self.submitMediaType(.video, url: destination)
+                }
+            } catch {
+                self.view?.showError(error)
+            }
+        }
+    }
+
+    func submitMediaType(_ type: MediaCommentType, url: URL, callback: @escaping (Error?) -> Void = { _ in }) {
+        guard let assignment = assignment, let userID = assignment.submission?.userID else { return }
+        let env = self.env
+        let mediaUploader = UploadMedia(type: type, url: url)
+        let uploading = SubmissionButtonAlertView.uploadingAlert(mediaUploader)
+        let reportError = { [weak self] (error: Error) -> Void in
+            let failure = UIAlertController(title: NSLocalizedString("Submission Failed", bundle: .student, comment: ""), message: error.localizedDescription, preferredStyle: .alert)
+            failure.addAction(UIAlertAction(title: NSLocalizedString("Retry", bundle: .student, comment: ""), style: .default) { _ in
+                self?.submitMediaType(type, url: url, callback: callback)
+            })
+            failure.addAction(UIAlertAction(title: NSLocalizedString("Cancel", bundle: .student, comment: ""), style: .cancel))
+            self?.view?.present(failure, animated: true, completion: nil)
+        }
+        let reportSuccess = { [weak self] in
+            let success = UIAlertController(title: NSLocalizedString("Success!", bundle: .student, comment: ""), message: nil, preferredStyle: .alert)
+            self?.view?.present(success, animated: true) {
+                DispatchQueue.main.asyncAfter(deadline: DispatchTime.now() + .milliseconds(400)) {
+                    success.dismiss(animated: true, completion: nil)
+                }
+            }
+        }
+        let doneUploading = { (error: Error?) -> Void in
+            DispatchQueue.main.async { uploading.dismiss(animated: true) {
+                if let error = error { reportError(error) }
+                reportSuccess()
```

Comment:
Should return after `reportError`, right? Otherwise you will report error and then report success...

Reply:
good catch